packages/trpc/server/trpc.ts

```diff
@@ -41,10 +43,41 @@ const isAdminMiddleware = t.middleware(({ ctx, next }) => {
     },
   });
 });
+interface IRateLimitOptions {
+  intervalInMs: number;
+  limit: number;
+}
+const isRateLimitedByUserIdMiddleware = ({ intervalInMs, limit }: IRateLimitOptions) =>
+  t.middleware(({ ctx, next }) => {
+    // validate user exists
+    if (!ctx.user) {
+      throw new TRPCError({ code: "UNAUTHORIZED" });
+    }
+
+    const { isRateLimited } = rateLimit({ intervalInMs }).check(limit, ctx.user.id.toString());
+
+    if (isRateLimited) {
+      throw new TRPCError({ code: "TOO_MANY_REQUESTS" });
+    }
+
+    return next({
+      ctx: {
+        // infers that `user` and `session` are non-nullable to downstream procedures
+        session: ctx.session,
+        user: ctx.user,
+      },
+    });
+  });
 
 export const router = t.router;
 export const mergeRouters = t.mergeRouters;
 export const middleware = t.middleware;
 export const publicProcedure = t.procedure.use(perfMiddleware);
 export const authedProcedure = t.procedure.use(perfMiddleware).use(isAuthedMiddleware);
+export const authedRateLimitedProcedure = ({ intervalInMs, limit }: IRateLimitOptions) =>
+  t.procedure
+    .use(perfMiddleware)
+    .use(isAuthedMiddleware)
+    .use(isRateLimitedByUserIdMiddleware({ intervalInMs, limit }));
```

Comment:
@zomars just inline the logic here if you wanna skip the extra check, it'll be inferred

Reply:
WDYM?

Reply:
Instead of having a standalone middleware, you can have an inline one

Reply:
Just a callback fn instead of the isRateLimitedThing

Reply:
I'm failing to understand how to use isAuthedMiddleware inline and keeping the ctx inference.

Reply:
It just won't be reusable outside of this